Refactor and cleanup scheduling queues for consistency, logging, and documentation

togettoyou opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

1. Refactor `unschedulablePods` into `unschedulablePodsQueuer` interface
   - Provides clearly separated methods: `add`, `update`, `delete`, `get`, `has`, `list`, `listPod`, `len`, and `clear`.
2. Moves `"Pod moved to an internal scheduling queue"` logging into queue `add` methods to ensure consistency.
3. Adds interface comments to `activeQueuer` and `SchedulingQueue`.

#### Which issue(s) this PR is related to:

#130977

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

togettoyou commented:
/cc @macsko

I will wait for PR [#132730](https://github.com/kubernetes/kubernetes/pull/132730) to be merged and then rebase this one accordingly.

togettoyou commented:
/retest

togettoyou commented:
@macsko  Do we have plans to continue moving this forward?

togettoyou commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1115,13 +1133,13 @@ func (p *PriorityQueue) moveAllToActiveOrBackoffQueue(logger klog.Logger, event
 
-	unschedulablePods := make([]*framework.QueuedPodInfo, 0, len(p.unschedulablePods.podInfoMap))
-	for _, pInfo := range p.unschedulablePods.podInfoMap {
+	podInfoList := make([]*framework.QueuedPodInfo, 0, p.unschedulablePods.len())
```
Oh, I can change this back.

togettoyou commented on pkg/scheduler/backend/queue/unschedulable_pods.go:
```diff
@@ -72,14 +104,39 @@ func (u *unschedulablePods) delete(pod *v1.Pod, gated bool) {
+
+// list returns all QueuedPodInfo objects in the unschedulable podInfoMap.
+// The returned map reflects the current internal state and must be treated as read-only by the caller.
+func (u *unschedulablePods) list() map[string]*framework.QueuedPodInfo {
```
Since `map` is a keyword in Go, Among alternatives like `all`, `items`, `asMap`, or `entries`, which one do you think fits better here?

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -114,13 +116,27 @@ type SchedulingQueue interface {
 	// Done must be called for pod returned by Pop. This allows the queue to
 	// keep track of which pods are currently being processed.
 	Done(types.UID)
+	// Update updates a pod in the active or backoff queue if present. Otherwise, it removes
```
How about rephrasing this:
```
    // Update updates pod in queues.
    // If pod is present in active or backoff queue, it gets updated with the new value.
    // If pod is in unschedulable queue and it is updated in a way that it may become schedulable, it gets moved from unschedulable to active queue.
	// If pod is not present in any of the queues, it gets added to active queue.
```

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -114,13 +116,27 @@ type SchedulingQueue interface {
+	// become schedulable and adds the updated one to the active queue.
+	// If pod is not present in any of the queues, it is added to the active queue.
 	Update(logger klog.Logger, oldPod, newPod *v1.Pod)
+	// Delete deletes the item from either of the two queues. It assumes the pod is
```
there are 3 queues (active, backoff, unschedulable)

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -114,13 +116,27 @@ type SchedulingQueue interface {
+	// only in one queue.
 	Delete(pod *v1.Pod)
+	// MoveAllToActiveOrBackoffQueue moves all pods from unschedulablePods to activeQ or backoffQ.
+	// This function adds all pods and then signals the condition variable to ensure that
```
```suggestion
	// This function adds all pods and afterwards signals the condition variable to ensure that
```

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -114,13 +116,27 @@ type SchedulingQueue interface {
 	// (e.g., filter Pods based on added/updated Node's capacity, etc.)
 	// We know currently some do, but we'll eventually remove them in favor of the scheduling queue hint.
 	MoveAllToActiveOrBackoffQueue(logger klog.Logger, event fwk.ClusterEvent, oldObj, newObj interface{}, preCheck PreEnqueueCheck)
+	// AssignedPodAdded is called when a bound pod is added. Creation of this pod
```
bound pod is added where? to the cluster? to the queue?

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -114,13 +116,27 @@ type SchedulingQueue interface {
+	// AssignedPodAdded is called when a bound pod is added. Creation of this pod
+	// may make pending pods with matching affinity terms schedulable.
 	AssignedPodAdded(logger klog.Logger, pod *v1.Pod)
+	// AssignedPodUpdated is called when a bound pod is updated. Change of labels
```
same as above

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -134,12 +150,20 @@ type SchedulingQueue interface {
 
 	// The following functions are supposed to be used only for testing or debugging.
+
+	// GetPod searches for a pod in the activeQ, backoffQ, and unschedulablePods.
```
Can you please describe return values as well?

togettoyou replied:
```go
	// GetPod searches for a pod in the activeQ, backoffQ, and unschedulablePods.
	// returning the pod info and true if found, or nil and false otherwise.
	GetPod(name, namespace string) (*framework.QueuedPodInfo, bool)
```

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -134,12 +150,20 @@ type SchedulingQueue interface {
+	// GetPod searches for a pod in the activeQ, backoffQ, and unschedulablePods.
 	GetPod(name, namespace string) (*framework.QueuedPodInfo, bool)
+	// PendingPods returns all the pending pods in the queue; accompanied by a debugging string
+	// recording showing the number of pods in each queue respectively.
```
```suggestion
	// recording the number of pods in each queue respectively.
```

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -134,12 +150,20 @@ type SchedulingQueue interface {
 	PodsInActiveQ() []*v1.Pod
 	// PodsInBackoffQ returns all the Pods in the backoffQ.
 	PodsInBackoffQ() []*v1.Pod
+	// UnschedulablePods returns all the pods in unschedulable state.
```
```suggestion
	// UnschedulablePods returns all the pods in unschedulable queue.
```

ania-borowiec reviewed (commented):
Thank you for this PR, this is a valuable change!
So far I've reviewed the doc comment updates, I'll proceed to reviewing actual code changes next.

The doc comments look good, good work! I added multiple comments, but most of them are just about typos or phrasing, nothing big.

ania-borowiec commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1062,7 +1079,7 @@ func (p *PriorityQueue) Update(logger klog.Logger, oldPod, newPod *v1.Pod) {
 	// If the pod is in the unschedulable queue, updating it may make it schedulable.
-	if pInfo := p.unschedulablePods.get(newPod); pInfo != nil {
+	if pInfo, exists := p.unschedulablePods.get(newPod); exists {
 		_ = pInfo.Update(newPod)
```
@macsko not related to this PR, but should this result (potential error) be ignored?

ania-borowiec commented on pkg/scheduler/backend/queue/unschedulable_pods.go:
```diff
@@ -18,11 +18,36 @@ package queue
 
+// unschedulablePodsQueuer is a wrapper for unschedulableQ related operations.
+type unschedulablePodsQueuer interface {
+	// add adds the pInfo to unschedulable podInfoMap.
+	// The event should show which event triggered the addition and is used for the metric recording.
```
This comment mentions implementation details and thus belongs to struct, not interface. Interface comment needs to be a bit more high level.

Same for all comments here that mention podInfoMap

ania-borowiec commented on pkg/scheduler/backend/queue/unschedulable_pods.go:
```diff
@@ -18,11 +18,36 @@ package queue
+	entries() map[string]*framework.QueuedPodInfo
+	// list returns all Pods in the unschedulable podInfoMap.
+	list() []*v1.Pod
+	// len returns length of the unschedulablePods.
```
```suggestion
	// len returns length of the unschedulablePods queue.
```

ania-borowiec commented on pkg/scheduler/backend/queue/unschedulable_pods.go:
```diff
@@ -43,18 +68,23 @@ func newUnschedulablePods(unschedulableRecorder, gatedRecorder metrics.MetricRec
-func (u *unschedulablePods) addOrUpdate(pInfo *framework.QueuedPodInfo, event string) {
+func (u *unschedulablePods) add(logger klog.Logger, pInfo *framework.QueuedPodInfo, event string) {
 	podID := u.keyFunc(pInfo.Pod)
-	if _, exists := u.podInfoMap[podID]; !exists {
```
Why did you remove this check? Are you sure that the pod is not already there?

togettoyou commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -134,12 +150,20 @@ type SchedulingQueue interface {
+	// recording showing the number of pods in each queue respectively.
+	// This function is used for debugging purposes in the scheduler cache dumper and comparer.
 	PendingPods() ([]*v1.Pod, string)
+	// InFlightPods returns all inFlightPods.
```
```go
	// InFlightPods returns all pods that are currently being processed by the scheduler.
	// These include pods that have been popped from the queue and are in their scheduling
	// or binding cycle. Pods are removed from inFlightPods once their scheduling succeeds or fails.
	InFlightPods() []*v1.Pod
```

such?

togettoyou commented on pkg/scheduler/backend/queue/unschedulable_pods.go:
```diff
@@ -43,18 +68,23 @@ func newUnschedulablePods(unschedulableRecorder, gatedRecorder metrics.MetricRec
+	u.podInfoMap[podID] = pInfo
+}
+
+// update updates the QueuedPodInfo in unschedulable podInfoMap.
```
Similarly, for the places where `update` is called, the existence has already been confirmed.

togettoyou replied:
However, it’s uncertain whether this might cause unnecessary issues later. Should we revert it back to `addOrUpdate`?

togettoyou commented:
> Thank you @togettoyou for pinging us. At this point the reviewer capacity in sig-scheduling is still very limited (just as it was all 2025). Hence we are focused on the critical reviews and take care of the less urgent ones whenever we have spare cycles. We are working hard on extending the number of reviewers, which will hopefully mean more PRs reviewed, too, and relatively soon.
> 
> Thank you for your patience!

Totally understandable, and many thanks for all your hard work.

ania-borowiec commented:
/reopen
/remove-lifecycle rotten

togettoyou commented:
I'll get this fixed later.

togettoyou committed:
scheduler: clean up unschedulable entity queue

togettoyou commented:
Rebased and reworked this on top of latest master.

The old `unschedulablePods` cleanup now maps to the current `unschedulableEntities` code, so it should work with both pods and pod groups. I also went back through the earlier review comments and avoided the previous list-allocation / naming issues.

Main changes are still the same idea: split `addOrUpdate` into `add` and `update`, add `has` / `(entity, ok)` lookup, and keep the queue-move log inside `add`, consistent with activeQ/backoffQ.


@ania-borowiec @macsko 
Really appreciate your patience and the previous reviews.

togettoyou commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1029,7 +1033,7 @@ func (p *PriorityQueue) AddUnschedulablePodIfNotPresent(logger klog.Logger, pInf
 	// In this case, we try to requeue this Pod to activeQ/backoffQ.
 	queue := p.requeueEntityWithQueueingStrategy(logger, pInfo, schedulingHint, framework.ScheduleAttemptFailure)
-	logger.V(3).Info("Pod moved to an internal scheduling queue", "pod", klog.KObj(pod), "event", framework.ScheduleAttemptFailure, "queue", queue, "schedulingCycle", podSchedulingCycle, "hint", schedulingHint, "unschedulable plugins", rejectorPlugins)
+	logger.V(5).Info("Pod requeue decision after scheduling attempt", "pod", klog.KObj(pod), "event", framework.ScheduleAttemptFailure, "queue", queue, "schedulingCycle", podSchedulingCycle, "hint", schedulingHint, "unschedulable plugins", rejectorPlugins)
```
Renamed this log because the actual “moved to an internal scheduling queue” log is now emitted by the target queue’s `add()` method. This line is more about the requeue decision after a failed scheduling attempt, with extra context like `hint`, `schedulingCycle`, and rejector plugins.

Same for the pod group log below.

togettoyou commented:
/retest
